File: lib/mutations/AddFeed.ts
import { graphql, Environment } from "react-relay"
import { ROOT_ID } from "relay-runtime"
import { commitMutationAsync } from "./commitMutationAsync"

const mutation = graphql`
  mutation AddFeedMutation($input: AddFeedInput!) {
    addFeed(input: $input) {
      feedEdge {
        node {
          ...FeedCard_feed
        }
        cursor
      }
    }
  }
`

export async function addFeed(
  environment: Environment,
  url: string
): Promise<void> {
  const variables = { input: { url } }
  await commitMutationAsync(environment, {
    mutation,
    variables,
    configs: [
      {
        parentID: ROOT_ID,
        type: "RANGE_ADD",
        connectionInfo: [
          {
            key: "FeedList_allSubscribedFeeds",
            rangeBehavior: "append",
          },
        ],
        edgeName: "feedEdge",
      },
    ],
  })
}

File: ios/Courier/Courier/Mutations/AddFeed.swift
import Relay
import RelaySwiftUI
import CourierGenerated

private let mutation = graphql("""
mutation AddFeedMutation($input: AddFeedInput!) {
  addFeed(input: $input) {
    feedEdge {
      node {
        ...FeedRow_feed
      }
      cursor
    }
  }
}
""")

private let updater: SelectorStoreUpdater = { store, data in
    let handler = ConnectionHandler.default
    guard
        let viewer = store.root.getLinkedRecord("viewer"),
        var feeds = handler.getConnection(viewer, key: "FeedsSection_allFeeds"),
        let newEdge = store.getRootField("addFeed")?.getLinkedRecord("feedEdge")
    else {
        return
    }

    handler.insert(connection: &feeds, edge: newEdge, after: nil)
}

extension Mutation.Mutator where Operation == AddFeedMutation {
    func commit(url: String, completion: ((Result<AddFeedMutation.Data?, Error>) -> Void)? = nil) {
        commit(
            variables: .init(input: .init(url: url)),
            updater: updater,
            completion: completion
        )
    }
}
